src/VisualStudio/LiveShare/Impl/PreviewCodeActionsHandler.cs

```diff
@@ -0,0 +1,66 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.LanguageServer;
+using Microsoft.CodeAnalysis.LanguageServer.CustomProtocol;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.VisualStudio.LiveShare.LanguageServices;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.VisualStudio.LanguageServices.LiveShare
+{
+    [ExportLspRequestHandler(LiveShareConstants.RoslynContractName, RoslynMethods.CodeActionPreviewName)]
+    internal class PreviewCodeActionsHandler : CodeActionsHandlerBase, ILspRequestHandler<RunCodeActionParams, LSP.TextEdit[], Solution>
+    {
+        [ImportingConstructor]
+        public PreviewCodeActionsHandler(ICodeFixService codeFixService, ICodeRefactoringService codeRefactoringService)
+            : base(codeFixService, codeRefactoringService)
+        {
+        }
+
+        public async Task<LSP.TextEdit[]> HandleAsync(RunCodeActionParams request, RequestContext<Solution> requestContext, CancellationToken cancellationToken)
+        {
+            var edits = ArrayBuilder<LSP.TextEdit>.GetInstance();
+            var solution = requestContext.Context;
+            var codeActions = await GetCodeActionsAsync(solution,
+                                                        request.TextDocument.Uri,
+                                                        request.Range,
+                                                        cancellationToken).ConfigureAwait(false);
+
+            var actionToRun = codeActions?.FirstOrDefault(a => a.Title == request.Title);
+
+            if (actionToRun != null)
+            {
+                var operations = await actionToRun.GetOperationsAsync(cancellationToken).ConfigureAwait(false);
+                var applyChangesOperation = operations.OfType<ApplyChangesOperation>().FirstOrDefault();
+
+                var document = solution.GetDocumentFromURI(request.TextDocument.Uri);
+                var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+                if (applyChangesOperation != null && document != null)
```

Comment:
I first thought there might be switch to UI thread, but there is none. so can we just share one from LSP side?

Reply:
In the last revision you mentioned I should move it to the liveshare layer since it was a custom method (not lsp)

Reply:
In the last revision you mentioned I should move it to the liveshare layer since it was a custom method (not lsp)